### ISSUE TITLE
Add Dictionary Support

### ISSUE THREAD

**Issue by nightblade9** [nightblade9/haxesharp#3]
# Motivation
As a C# developer, I find the `Map` class difficult to use (compiler errors about the `Map` sub-type are confusing; checking for specific keys/values is difficult; fetching all values is impossible).

I would like a `Dictionary` class that resembles what C# provides.

# MVP

To support this use-case, at a minimum, we need to implement:

- Array accessor to get and set values. [Haxe supports this](https://haxe.org/manual/types-abstract-array-access.html) via the `@:arrayAccess` annotation.
- `containsKey(key)` method
- `count` property
- `keys` property to get all keys
- `values` property to get all values
- Constructor with no parameters
- Constructor that takes a `Map` as a parameter and uses those values
- `toMap` method that returns a `Map`

Later, we can add support for other convenience methods as needed:

- `add(key, value)` method
- `clear` method
- `containsValue(value)` method
- `remove(key)` method

Each method can be individually added (with appropriate unit tests + docs).

# Internal Design

Haxe provides a `Map` class with three implementations (depending on the key type). Internally, we should store things in a `Map` instance. (This also simplifies the constructor and `toMap` methods.)

We can pass the type parameters along to the `Map` instance constructor, and everything should Just Work.

The only real difficulty is how to efficiently retrieve all the values. We can use two `Maps` (one with `value => key` mappings) or an array with values, but this effectively doubles the required memory and forces us to keep everything synchronized. If retrieving values is an infrequent operation, it might be okay to just iterate over all keys and retrieve all values in linear time. (We can ship the MVP with this.)

# Next Steps

I would approach this by implementing the following:

- Basic `Dictionary` class in `haxesharp.collections` with a constructor and array-access to get/set values
- Unit tests to cover the [four types of `Map` that Haxe abstracts over](https://haxe.org/manual/std-Map.html):
  - Unit tests with `string` keys
  - Unit tests with `int` keys
  - Unit tests with `enum` values for keys
  - Unit tests with object values for keys

Once that's done, the rest of the methods should be easy to selectively implement.

@bsinky is this something you would like to work on?

**Comment by bsinky** [comment]
Yes, this sounds interesting, I can work on this.  Thanks for the robust write-up.

> Each method can be individually added (with appropriate unit tests + docs).

By this, do you imagine the initial pull request to just add the Dictionary class with a constructor and array-access get/set methods, and additional (later) pull requests would add more C# Dictionary methods (`containsKey`, `count`, `keys`, etc.) ?

**Comment by nightblade9** [comment]
> Thanks for the robust write-up.

This is what we agreed in #1 about "more direction/coaching!"

> By this, do you imagine the initial pull request to just add the Dictionary class with a constructor and array-access get/set methods, and additional (later) pull requests would add more C# Dictionary methods (containsKey, count, keys, etc.) ?

Yes, precisely. This can be done in a series of commits and/or PRs (as you see fit). 

I think the "best practices" with Git recommend branching per-feature or per-change, although admittedly I don't do that much. Feel free to use whatever works for you.

**Comment by bsinky** [comment]
So, I have hit a stumbling block that surprised me.  I thought I would be able to do something this:

```haxe
class Dictionary<TKey, TValue>
{
    private var _map:Map<TKey, TValue>;

    public function new(?map:Map<TKey, TValue>)
    {
        if (map != null)
            _map = map;
        else
            _map = new Map<TKey, TValue>();
    }
}
```
But trying to compile this (by running unit tests) fails like so:

```
HaxeWrapper.hx:73: src/haxesharp/collections/Dictionary.hx:13: characters 19-42 : Abstract Map has no @:to function that accepts IMap<haxesharp.collections.Dictionary.TKey, haxesharp.collections.Dictionary.TValue>
```
I think I understand the error, `Map` is an `abstract` and doesn't declare a `@:to` method for casting to `IMap`.  I don't really get why a `@:to` `IMap` method is trying to be accessed here.

I see it as, I'm trying to assign from a `Map<TKey,TValue>` reference (which will be resolved to one of the 4 concrete implementations, `StringMap`, `IntMap`, etc. at compile-time), to another `Map<TKey,TValue>` reference, so it should be acceptable? But, obviously this is not the case.

**Comment by nightblade9** [comment]
I don't know the answer, but I figured "something" out. Not sure if it's Good Enough -- as long as it passes all the unit tests, I think we're okay. (For the first version, we're okay with poor performance, because we're solving problems like this one.)

```
class Dictionary<K : {}, V>
{
    private var _map:ObjectMap<K, V>;

    public function new(?map:ObjectMap<K, V>)
    {
        if (map != null)
            _map = map;
        else
            _map = new ObjectMap<K, V>();
    }
}
```

You can play with a minimal example [here](https://try.haxe.org/#3258C). The two things to note:

- I switched to an `ObjectMap` instead of a `Map`. This is to get around the compilation error. You can try asking on the Haxe mailing list if there's a better way to do this.
- I added the `K : {}` constraint to satisfy another compiler error. I figured it out by looking at the [ObjectMap source code](https://github.com/HaxeFoundation/haxe/blob/development/std/haxe/ds/ObjectMap.hx). I reached that by looking up `Map` on api.haxe.org and then clicking View Source when I didn't see any mention of constraints.

This *may* work, and it may or may not perform poorly, but I think this is enough to unblock you. Please ask on the Haxe mailing list about the compiler error you reported here; you may also want to ask if there's any downside to always using `ObjectMap` and not `Map` (I assume some sort of performance loss, but if it's 10-20% CPU or RAM then we may be okay with that).

**Comment by bsinky** [comment]
It seems like using `ObjectMap` isn't going to be an option either:

```haxe
var intDict = new Dictionary<Int, Int>();
```

...gives an error similar to `HaxeWrapper.hx:73: test/haxesharp/collections/DictionaryTest.hx:24: characters 31-57 : Int should be { }`

This also happens with `enum` types.

I'll keep poking around different ways to could implement this.  Once I send a message on the Haxe mailing list I'll link to the thread here.

Perhaps we could change the focus of this, and instead focus on creating a `MapExtensions` class of static extension methods.

The benefit of seeing `Dictionary` as the type would be lost, C# developers coming to Haxe would need to get used to `Map` and its sub-types instead (and deal with confusing error messages as you mentioned earlier).

**Comment by nightblade9** [comment]
Yeah, we would lose most of the benefits by switching to static extensions. This is exactly the type of craziness Haxe produces that I don't want to deal with.

Another option might be to (internally) wrap enums, ints, and other problematic types (strings?) into some sort of object. That would make some of the operations more complex (eg. getting something with a string key when we store it in an object with a string property), but it might be worth it. 

What do you think is the best way to proceed?

**Comment by bsinky** [comment]
I've reached out to the Haxe mailing list: https://groups.google.com/forum/#!topic/haxelang/UQ7oknlHyDA

I'm hopeful someone there will either be able to explain exactly what's going wrong here, and (ideally), provide a workaround that is both efficient and not-too-hacky.

In the meantime, I'll spend a little more time digging into `Map`, and `IMap` and its implementations.

**Comment by nightblade9** [comment]
[I just tested this again](https://try.haxe.org/#9AdE9). What you're missing is the constraint `K : {}` on the type for the key.

Some style things:
- Please call your type parameters `K` and `V`
- Please always always *always* use curly-braces for if statements. It prevents bugs.
- Please don't name member variables with an underscore. That's not a recommended style.

**Comment by bsinky** [comment]
I should have mentioned earlier, using the { } constraint and ObjectMap
works for Strings. But, if you try Dictionary <Int, String> in your
example, it will not compile.

On Oct 13, 2017 4:07 PM, "nightblade9" <notifications@github.com> wrote:

I just tested this again <https://try.haxe.org/#9AdE9>. What you're missing
is the constraint K : {} on the key.

**Comment by nightblade9** [comment]
That's probably worth mentioning in the forums post.

I wonder what `K : {}` really means/does?

**Comment by nightblade9** [comment]
Closed with #4. The current implementation is definitely efficient (uses `IntMap`, etc. as appropriate) but requires you to know about `@:abstract`, etc. Perhaps we can make an iteratively simpler version in the future.

Thanks for working through all the low-level Haxe implementation issues to get this figured out @bsinky .

For reference, #4 mentions a simpler implementation with an `ObjectMap` and wrapping all keys (ints, strings, enums, and objects) in an object class for a simpler, uniform API; but `IntMap`, etc. have very efficient implementations across targets.
